tests/integration/test_storage_s3_queue/test_parallel_inserts.py

```diff
@@ -0,0 +1,246 @@
+import logging
+import time
+import math
+import pytest
+
+from helpers.cluster import ClickHouseCluster
+from helpers.s3_queue_common import (
+    generate_random_files,
+    put_s3_file_content,
+    create_table,
+    create_mv,
+    generate_random_string,
+)
+
+
+@pytest.fixture(scope="module")
+def started_cluster():
+    try:
+        cluster = ClickHouseCluster(__file__)
+        cluster.add_instance(
+            "instance",
+            user_configs=["configs/users.xml"],
+            with_minio=True,
+            with_azurite=True,
+            with_zookeeper=True,
+            main_configs=[
+                "configs/zookeeper.xml",
+                "configs/s3queue_log.xml",
+                "configs/remote_servers.xml",
+            ],
+            stay_alive=True,
+        )
+
+        logging.info("Starting cluster...")
+        cluster.start()
+        logging.info("Cluster started")
+
+        yield cluster
+    finally:
+        cluster.shutdown()
+
+
+def run_with_retry(check_result, func, retries=100):
+    for _ in range(retries):
+        last = func()
+        if check_result(last):
+            return last
+        time.sleep(1)
+    raise RuntimeError(f"{last} did not match expectations in {retries} retries")
+
+
+@pytest.mark.parametrize("parallel_inserts", [0, 1])
```

Comment:
btw, it seems all s3queue tests apart from a few (those which test "commit limits") should work with parallel inserts enabled?

Reply:
I guess (haven't checked though)